### PR TITLE
Convert Http2Connection to Kotlin

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/internal/UtilKt.kt [hunk 1]
```diff
@@ -36,3 +40,37 @@ fun BufferedSource.readMedium(): Int {
...
+/** Run [block] until it either throws an [IOException] or completes. */
+inline fun ignoreIoExceptions(block: () -> Unit) {
+  try {
+    block()
+  } catch (_: IOException) {
+  }
+}
+
+/** Execute [block], setting the executing thread's name to [name] for the duration. */
+fun Executor.execute(name: String, block: () -> Unit) {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/UtilKt.kt]: `inline` + `crossinline`

File: okhttp/src/main/java/okhttp3/internal/UtilKt.kt [hunk 2]
```diff
@@ -36,3 +40,37 @@ fun BufferedSource.readMedium(): Int {
...
+  execute(object : NamedRunnable("%s", name) {
+    override fun execute() {
+      block()
+    }
+  })
+}
+
+/** Executes [block] unless this executor has been shutdown, in which case this does nothing. */
+fun Executor.tryExecute(name: String, block: () -> Unit) {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/UtilKt.kt]: `inline` + `crossinline`, re-use function above in implementation

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt [hunk 3]
```diff
@@ -0,0 +1,921 @@
...
+   * for zero [byteCount] is closing a flushed output stream.
+   */
+  @Throws(IOException::class)
+  fun writeData(
+    streamId: Int,
+    outFinished: Boolean,
+    buffer: Buffer?,
+    byteCount: Long
+  ) {
+    var byteCount = byteCount
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt]: Move down directly above `while`?

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt [hunk 4]
```diff
@@ -0,0 +1,921 @@
...
+    cause: IOException?
+  ) {
+    assert(!Thread.holdsLock(this))
+    ignoreIoExceptions {
+      shutdown(connectionCode)
+    }
+
+    var streamsToClose: Array<Http2Stream>? = null
+    synchronized(this) {
+      if (!streams.isEmpty()) {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt]:
```suggestion
      if (streams.isNotEmpty()) {
```

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt [hunk 5]
```diff
@@ -0,0 +1,921 @@
...
+        streamsToClose = streams.values.toTypedArray()
+        streams.clear()
+      }
+    }
+
+    if (streamsToClose != null) {
+      for (stream in streamsToClose!!) {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt]: ugh. could be `streamsToClose?.forEach { stream -> .. }` if you want

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt [hunk 6]
```diff
@@ -0,0 +1,921 @@
...
+  internal fun pushDataLater(
+    streamId: Int,
+    source: BufferedSource,
+    byteCount: Int,
+    inFinished: Boolean
+  ) {
+    val buffer = Buffer()
+    source.require(byteCount.toLong()) // Eagerly read the frame before firing client thread.
+    source.read(buffer, byteCount.toLong())
+    if (buffer.size != byteCount.toLong()) {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt]: Er... impossible?

File: okhttp/src/main/java/okhttp3/internal/UtilKt.kt (hunk 1 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/UtilKt.kt]: Done!

File: okhttp/src/main/java/okhttp3/internal/UtilKt.kt (hunk 2 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/UtilKt.kt]: Done!

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt (hunk 3 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt]: Done!

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt (hunk 4 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt]: Done!

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt (hunk 5 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt]: Done.

Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt]: Shouldn’t be a problem with contracts?

File: okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt (hunk 6 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt]: Impossible.

Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/http2/Http2Connection.kt]: (removed!)

**Pull request merged**